File: Binary_Search/Ruby/jcla1/binary_search.rb
def binary_search(arr, target)
    if arr.length < 1
        return -1
    end

    middle = arr.length / 2
    middle_value = arr[middle]

    if middle_value > target
        binary_search(arr[0, middle], target)
    elsif middle_value < target
        v = binary_search(arr[(middle+1)..-1], target)
        v < 0 ? v : (v + middle + 1)
    else
        middle
    end
end

File: Binary_Search/Python/jcla1/binary_search.py
# Expects the array to be sorted
def binary_search(arr, target):
    if len(arr) < 1:
        return -1

    middle = int(len(arr) / 2)
    middle_value = arr[middle]

    if middle_value > target:
        return binary_search(arr[:middle], target)
    elif middle_value < target:
        v = binary_search(arr[middle + 1:], target)
        # since we're jumping forward in the slice (arr[middle+1:])
        # we need to compensate that by adding the jump size
        if v < 0:
            return v
        else:
            return middle + v + 1

    return middle
